### ISSUE TITLE
[not an issue] Alternative approach to lazy-loading

### ISSUE THREAD

**Issue by MiguelMadero** [duizendnegen/ember-cli-lazy-load#9]
_Note:_ note really an "issue", just opening a thread for discussion. 

Hi @B-Stefan and @duizendnegen I have been working on a POC to build and lazy-load `packages`. I'm impressed by how similar both approaches are. Some things that are different from mine that might be worth mentioning are: 
- Each app is in its own folder. This is aligned better with addons so ease the migration path, but it also reduces the need to use glob to specify what is in each bundle. 
- Instead of relying on a mixin to turn routes into lazy-loaded routes, I have a catchAll route that is used to load the new bundle, add the route definitions to the router and finally retry the original transition, see [routes/catch-all#redirect](https://github.com/MiguelMadero/ember-cli-packages-demo/blob/master/app/routes/catch-all.js#13)

From this project, I really like the instructions to set this up, is super straight forward and easy to reason about. The other is still a bit tangled, but I plan to extract some of the parts into an add-on to make the setup a bit simpler. 

You can have a look at a [demo app](https://github.com/MiguelMadero/ember-cli-packages-demo), I would love your feedback. I know this project hasn't been active for a couple of months, but this is a problem that I need to solve really soon and I have time to throw at it.

**Comment by duizendnegen** [comment]
Good approach, very interesting. Having each app in its own folder is good for organisation, think the same could be achieved with pods though. I like the catch-all route, should work well.

Let's try and see how we can go ahead and push both projects forward together - doesn't make sense to have these split efforts.

**Comment by MiguelMadero** [comment]
> Let's try and see how we can go ahead and push both projects forward together

Any suggestions? There are a few outstandings tasks that I wanted to tackle. Maybe once those three are done it would be clearer how to get the best of both. 

Below are the outstanding items: 

[ ] Lazy load CSS
[ ] Make sure {{link-to}} works or use somethign that avoid the dependency on Route#serialize
[ ] Extract into an add-on

Do you have this approach in a prod apps? Or just have it as a POC? I'm wondering about how it works in the wild. I used the packaging app of this on a really large app and last year, I used was building an Ember 1.5 app in Grunt that used the catch-all approach to lazy load, but that relied on calling `Router.map` multiple times, but that stopped working in 1.10. This is the first time I'm doing this with CLI and the using the router DSL.

**Comment by duizendnegen** [comment]
I do not have this in prod at the moment, actually running `gzip` over our assets made them small enough - only issue is JS interpretation times on mobile still.

As for concrete suggestions: I'd hate to see two projects that do the same thing - I think the goals of our projects are identical. Ideally I'd want to merge and deprecate one of them.
However, I wouldn't want to make too much hassle out of it, with `ember-engines` also landing some time soon.

As per your outstanding items:
- Lazy loading CSS can be done overriding the CSS output tree, you'll have to be very explicit though as to which packages do so. Also, keep in mind what are the wins here: interpretation and download times of CSS where less than 10% of the JS times in my case
- `{{link-to}}` works out of the box with the Mixin as is used in this project
- this project is an addon. You can `npm install` it, and import the mixins where needed.

**Comment by MiguelMadero** [comment]
re: CSS. The build part works fine. It is just the loading part that is missing, but that is an easy part. 

re: `{{link-to}}`. I'll have a look to see how it works. The scenario that failed for me is when using `link-to` in combination with a model, Ember will look for a route to call `serialize`, which has to be available. See more on [ember-cli/rfc#](https://github.com/emberjs/rfcs/pull/120). 

I think we are both on the same page, that eventually both projects will be deprecated once the engines dream is there. 

re: JS interpretation in mobile. That's a big problem, even on desktop computers with really large apps I saw delays of ~3 seconds just interpreting the JS. That's all sync, which leads to a horrible UX. In the past, prior to lazy loading, I found that loading the scripts in the way both projects are doing (adding a script tag from JS) led to a huge difference since scripts are executed asynchronously in most major browsers and in Firefox even compiled on a background thread. For apps not using this technique, adding an async flag to the scrip tag should have the same effect.
